### PR TITLE
lxd: add support for interim Ubuntu releases

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----
#### Overview
This PR adds support for interim and devel Ubuntu releases with LXD.

Broken into 6 separate commits to speed up review time.

#### Details

Previously, craft-providers maintained a list of valid LXD images with a simple enum.  Now, the `remotes.py` module has a new dataclass called `RemoteImage` and a dictionary of valid `RemoteImages` to track what images are supported.

It is designed in such a way that users can add custom remotes and we'll be able to add support for other bases (like a centos base) without breaking changes.

#### Usage

For applications using the `LxdProvider` class (snapcraft, charmcraft, and rockcraft), there are no changes in behavior or usage.  You can simply pass in the new release names:
```python
class BuilddBaseAlias(enum.Enum):
    XENIAL = "16.04"
    BIONIC = "18.04"
    FOCAL = "20.04"
    JAMMY = "22.04"
    KINETIC = "22.10"
    LUNAR = "23.04"
```

For other applications (lpcraft :eyes:), you'll need to use the replace the old `configure_buildd_image_remote()` function with the new `get_remote_image()` function.

Note - the old function `configure_build_image_remote()` still exists and its behavior is unchanged, except for a deprecation notice.

#### Example usage

```python
remote_image = lxd.get_remote_image(BuilddBaseAlias.KINETIC.value)

base_configuration = bases.BuilddBase(alias=BuildBaseAlias.KINETIC)

instance = lxd.launch(
    name="my-instance",
    base_configuration=base_configuration,
    image_name=remote_image.image_name,
    image_remote=remote_image.remote_name,
)
```

#### TODO
- [x] break into smaller commits

#### Sources

- [Provide build base support for interim releases #190](https://github.com/canonical/craft-providers/issues/190)
- [Support new series of ubuntu including interim releases ](https://bugs.launchpad.net/lpcraft/+bug/1999694)

(CRAFT-1601)

### CODE REVIEW

**Comment by jugmac00** [comment]
From lpcraft side - it works like a charm! Thanks a lot!

File: craft_providers/lxd/remotes.py [hunk 1]
```diff
@@ -15,50 +16,165 @@
...
+    is_stable: bool
+
+    def add_remote(self, lxc: LXC = LXC()) -> None:
+        """Add the LXD remote for an image.
+
+        If the remote already exists, it will not be re-added.
+
+        :param lxc: LXC client.
+        """
+        if self.remote_name in lxc.remote_list():
```
Comment (lengau) [inline craft_providers/lxd/remotes.py:76]: Theoretically this means the user could modify the remote and cause problems (images not available, wrong images coming, etc.). It might be worth checking the remote address in that case and providing a warning if it differs. (Or maybe that's too much of an edge case to care about considering that this is the same logic as previously existed.)

File: craft_providers/lxd/remotes.py [hunk 2]
```diff
@@ -15,50 +16,165 @@
...
+    """Configure the default buildd image remote.
+
+    This is a deprecated function to maintain the existing API. It will be
+    removed with the release of craft-providers 2.0.
+
+    :param lxc: LXC client.
+
+    :returns: Name of remote to pass to launcher.
+    """
+    logger.warning(
```
Comment (lengau) [inline craft_providers/lxd/remotes.py:173]: I take it we don't use `warnings.warn` and `DeprecationWarning` here because it's filtered out by default?

File: tests/integration/lxd/test_lxd_provider.py [hunk 3]
```diff
@@ -40,16 +42,26 @@ def test_create_environment(installed_lxd, instance_name):
...
-def test_launched_environment(installed_lxd, instance_name, tmp_path):
+@pytest.mark.parametrize(
+    "alias",
+    [
+        BuilddBaseAlias.BIONIC,
+        BuilddBaseAlias.FOCAL,
+        BuilddBaseAlias.JAMMY,
+        BuilddBaseAlias.KINETIC,
+        BuilddBaseAlias.LUNAR,
+    ],
```
Comment (lengau) [inline tests/integration/lxd/test_lxd_provider.py:53]: Are we intentionally not testing on Xenial?

If we intend to test on all supported base aliases it may be better to do:

```suggestion
    "alias", list(BuilddBaseAlias),
```

so the tests automatically catch any new bases added.

File: craft_providers/lxd/remotes.py (hunk 2 above)
Comment (mr-cal) [reply craft_providers/lxd/remotes.py:173]: That is what I surmised.  I am open to suggestions for a better approach to deprecations.

File: tests/integration/lxd/test_lxd_provider.py (hunk 3 above)
Comment (mr-cal) [reply tests/integration/lxd/test_lxd_provider.py:53]: `xenial` was conspicuously missing and untested in craft-providers (except one integration test).  It would raise an exception when trying to launch it using the Provider classes.

It's still available for LXD, but not for Multipass.  `lpcraft` still uses `xenial`.  

I'll add support for it and use your suggestion too!

File: craft_providers/lxd/remotes.py (hunk 1 above)
Comment (mr-cal) [reply craft_providers/lxd/remotes.py:76]: You're right - it's a good check and an edge case.  I'll add a TODO for now

File: tests/integration/lxd/test_lxd_provider.py (hunk 3 above)
Comment (mr-cal) [reply tests/integration/lxd/test_lxd_provider.py]: After adding `xenial`, I'm seeing failures related to the version of `snapd` being too old.

File: craft_providers/lxd/remotes.py [hunk 4]
```diff
@@ -58,6 +58,40 @@ class RemoteImage:
     remote_protocol: ProtocolType
     is_stable: bool
 
+    def add_remote(self, lxc: LXC = LXC()) -> None:
```
Comment (tigarmo) [inline craft_providers/lxd/remotes.py:69]: out of curiosity, what's the semantics of this default value? When would we use that?

File: craft_providers/lxd/lxd_provider.py [hunk 5]
```diff
@@ -120,14 +114,26 @@ def launched_environment(
 
         :raises LXDError: if instance cannot be configured and launched
         """
-        image_remote = configure_buildd_image_remote()
+        image = get_remote_image(build_base)
+        image.add_remote()
+
+        # we can't guarantee daily and devel images, so explicitly warn the user
+        if not image.is_stable:
+            warnings.warn(
```
Comment (tigarmo) [inline craft_providers/lxd/lxd_provider.py:122]: I thought `warnings.warn()` was meant more for devs, and not end-users

File: craft_providers/lxd/remotes.py (hunk 2 above)
Comment (mr-cal) [reply craft_providers/lxd/remotes.py]: Updated to DeprecationWarning

File: craft_providers/lxd/lxd_provider.py (hunk 5 above)
Comment (mr-cal) [reply craft_providers/lxd/lxd_provider.py:122]: I believe @sergiusens requested this message be emitted with the `warnings` module so that it cannot be silenced or filtered out.

File: craft_providers/lxd/remotes.py (hunk 4 above)
Comment (mr-cal) [reply craft_providers/lxd/remotes.py]: It's a great question and I don't have a good answer.

I've kept it because it was already there in the `configure_buildd_image_remote(lxc: LXC = LXC())` function.

I think this would be a good opportunity to remove the default value.  Done!

File: tests/integration/lxd/test_lxd_provider.py (hunk 3 above)
Comment (lengau) [reply tests/integration/lxd/test_lxd_provider.py:53]: Welp, now we know! Thank you!

File: craft_providers/lxd/lxd_provider.py (hunk 5 above)
Comment (sergiusens) [reply craft_providers/lxd/lxd_provider.py:122]: My thought process was that we were enabling the devel images early and its used would be something an application would enable, not a user through configuration. I would for example slightly change the API and add an allow_unstable=False param here and raise your warning as an exception.

File: craft_providers/bases/buildd.py [hunk 6]
```diff
@@ -144,6 +144,8 @@ class BuilddBaseAlias(enum.Enum):
     BIONIC = "18.04"
     FOCAL = "20.04"
     JAMMY = "22.04"
+    KINETIC = "22.10"
+    LUNAR = "23.04"
```
Comment (sergiusens) [inline craft_providers/bases/buildd.py:148]: There's a devel alias in the buildd-daily I would like to see here
image

Comment (mr-cal) [reply craft_providers/bases/buildd.py:148]: This may be out of scope of this PR.  Adding the `DEVEL = "devel"` alias launches the image fine, but raises an error during setup because we are strictly matching `/etc/os-release` against the alias:

```
Incompatible base detected: Expected OS version 'devel', found '23.04'.
```

File: craft_providers/lxd/lxd_provider.py (hunk 5 above)
Comment (mr-cal) [reply craft_providers/lxd/lxd_provider.py]: As discussed offline, I changed this to an exception and made the `is_stable` property implicit.

**Review by sergiusens — APPROVED** [review]

**Comment by jugmac00** [comment]
Quick heads-up - we faced a (short) [production regression](https://bugs.launchpad.net/lpcraft/+bug/2009954) after deploying these changes, as it looks like that after my review some more changes were made, especially the work around `add_remote`, which the example section does not cover.

As I have tried the previous state of that branch, my local lxc installation was already configured with the buildd remote (`craft-com.ubuntu.cloud-buildd`), but in production that failed.

Also, the changelog for 1.8.0 is missing, see https://github.com/canonical/craft-providers/issues/233 and https://github.com/canonical/craft-providers/blob/c6e624516dbb76ff60abfe9a2459bffe288d5e0e/docs/changelog.rst

Thanks to my colleague @lgp171188 we noticed that issue rather quickly, we rolled it back, and I was able to provide a fix for lpcraft.

Anyway, it works now, thanks a lot for your work!

P.S.: And looks like I need some more isolated test environment.
